gemrb/core/Scriptable/Actor.cpp

```diff
@@ -4927,59 +4927,46 @@ void Actor::Turn(Scriptable *cleric, ieDword turnlevel)
 	}
 
 	//determine if we see the cleric (distance)
-	if (!CanSee(cleric, this, true, GA_NO_DEAD)) {
+	if (!CanSee(sender, this, true, GA_NO_DEAD)) {
 		return;
 	}
 
-	const Actor* cleric2 = Scriptable::As<Actor>(cleric);
-	if (cleric2 && GameScript::ID_Alignment(cleric2, AL_EVIL)) {
+	const Actor* caster = Scriptable::As<Actor>(sender);
+	if (caster && GameScript::ID_Alignment(caster, AL_EVIL)) {
 		evilcleric = true;
 	}
 
-	//a little adjustment of the level to get a slight randomness on who is turned
-	unsigned int level = GetXPLevel(true)-(GetGlobalID()&3);
+	int casterTurnLevel = turnlevel - 1 + core->Roll(1, 4, 0);
+	int targetTurnLevel = GetStat(IE_GENERAL) == GEN_UNDEAD ? GetXPLevel(true) : evilcleric ? GetPaladinLevel() : 0;
 
-	//this is safely hardcoded i guess
-	if (Modified[IE_GENERAL]!=GEN_UNDEAD) {
-		level = GetPaladinLevel();
-		if (evilcleric && level) {
-			AddTrigger(TriggerEntry(trigger_turnedby, cleric->GetGlobalID()));
-			if (turnlevel >= level + turnDeathLevelMod) {
-				if (gamedata->Exists("panic", IE_SPL_CLASS_ID)) {
-					core->ApplySpell(ResRef("panic"), this, cleric, level);
-				} else {
-					Log(DEBUG, "Actor", "Panic from turning!");
-					Panic(cleric, PANIC_RUNAWAY);
-				}
-			}
-		}
+	if (!targetTurnLevel) {
 		return;
 	}
 
 	//determine alignment (if equals, then no turning)
 
-	AddTrigger(TriggerEntry(trigger_turnedby, cleric->GetGlobalID()));
+	AddTrigger(TriggerEntry(trigger_turnedby, sender->GetGlobalID()));
 
 	//determine panic or destruction/control
 	//we get the modified level
-	if (turnlevel >= level + turnDeathLevelMod) {
+	if (casterTurnLevel >= targetTurnLevel + turnDeathLevelMod) {
 		if (evilcleric) {
 			Effect* fx = EffectQueue::CreateEffect(control_creature_ref, GEN_UNDEAD, 3, FX_DURATION_INSTANT_LIMITED);
 			if (!fx) {
 				fx = EffectQueue::CreateEffect(control_undead_ref, GEN_UNDEAD, 3, FX_DURATION_INSTANT_LIMITED);
 			}
 			if (fx) {
-				fx->Duration = core->Time.round_sec;
+				fx->Duration = core->Time.round_sec * caster->GetXPLevel(true);
```

Comment:
Nice catch, how did you get this value? The bg wiki is not the best resource, claiming a fixed 10 rounds, but the testing could have easily been bad (in the sense of testing the wrong variables).

Reply:
> Nice catch, how did you get this value? The bg wiki is not the best resource, claiming a fixed 10 rounds, but the testing could have easily been bad (in the sense of testing the wrong variables).

I'm not sure about this point, but the rest should be correct. Before the edits, turn undead threw one projectile at the caster himself.

Reply:
Done revert duration

Reply:
It's probably still wrong, since I doubt people would report 10 rounds if it was only 1. So we had a bug before. Just needs some testing to see if it scaled with level (if so, it should probably only take the cleric/paladin level into account, so eg. fighter/clerics don't get unjust power).

Reply:
Yes, I'll try to test it